### PR TITLE
Remove listeners in UndoAction and RedoAction for memory efficiency

### PR DESCRIPTION
**Pull request by melisolmez**
Fix https://github.com/JabRef/jabref/issues/11809 
Fix https://github.com/JabRef/jabref/issues/11837

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by koppor** [comment]
Checkstyle says

> Error: eckstyle] [ERROR] /home/runner/work/jabref/jabref/src/main/java/org/jabref/gui/undo/RedoAction.java:8:1: Wrong order for 'javax.swing.undo.CannotRedoException' import. [ImportOrder]

File: src/main/java/org/jabref/gui/undo/RedoAction.java [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
 import org.jabref.gui.actions.SimpleCommand;
 import org.jabref.logic.l10n.Localization;
 
 /**
- * @implNote See also {@link UndoAction}
+ * @implNote
+ * See
+ * also
+ * {@link
+ * UndoAction}
```
Comment (subhramit) [inline src/main/java/org/jabref/gui/undo/RedoAction.java:22]: Revert this

File: src/main/java/org/jabref/gui/undo/RedoAction.java [hunk 2]
```diff
@@ -34,7 +46,8 @@ public void execute() {
         try {
             libraryTab.getUndoManager().redo();
             dialogService.notify(Localization.lang("Redo"));
-        } catch (CannotRedoException ex) {
+        } catch (
+                CannotRedoException ex) {
```
Comment (subhramit) [inline src/main/java/org/jabref/gui/undo/RedoAction.java:50]: Please also fix the catch indent (not your fault, IntelliJ does this by default when we format the document).

**Review by subhramit — CHANGES_REQUESTED** [review]
Minor formatting changes requested. Otherwise LGTM.
Thank you for your contribution!

P.S. Do mark the mandatory checks that you have completed in the PR description when you file one.

File: src/main/java/org/jabref/gui/undo/RedoAction.java [hunk 3]
```diff
@@ -21,11 +25,15 @@ public RedoAction(Supplier<LibraryTab> tabSupplier, DialogService dialogService,
         this.tabSupplier = tabSupplier;
         this.dialogService = dialogService;
 
-        // TODO: The old listener should be removed. Otherwise, memory consumption will increase.
-        stateManager.activeTabProperty().addListener((observable, oldValue, activeLibraryTab) -> {
+        ChangeListener<Optional<LibraryTab>> listener = (observable, oldValue, activeLibraryTab) -> {
             activeLibraryTab.ifPresent(libraryTab ->
                     this.executable.bind(libraryTab.getUndoManager().getRedoableProperty()));
-        });
+
```
Comment (koppor) [inline src/main/java/org/jabref/gui/undo/RedoAction.java:31]: This empty line is not necessary, is it?

**Comment by koppor** [comment]
@subhramit We should test this. This PR implements both suggestions of you (https://github.com/JabRef/jabref/issues/11809#issuecomment-2367033790). I think, one needs to open 5 to 10 libraries, switch tabs back and forth, modifying the library and see if undo/redo is enabled properly.

@melisolmez For consistency reasons, this patch also nees to be ported to the UndoAction. Undo is the twin of Redo. --> https://github.com/JabRef/jabref/blob/8412e651b671a427c828924ce5698eb4c32f269d/src/main/java/org/jabref/gui/undo/UndoAction.java

File: src/main/java/org/jabref/gui/undo/RedoAction.java (hunk 3 above)
Comment (melisolmez) [reply src/main/java/org/jabref/gui/undo/RedoAction.java:31]: actually this empty line is for exists to separate lines 30 and 32 from each other. But I can remove.

**Comment by melisolmez** [comment]
> @subhramit We should test this. This PR implements both suggestions of you ([#11809 (comment)](https://github.com/JabRef/jabref/issues/11809#issuecomment-2367033790)). I think, one needs to open 5 to 10 libraries, switch tabs back and forth, modifying the library and see if undo/redo is enabled properly.
> 
> @melisolmez For consistency reasons, this patch also nees to be ported to the UndoAction. Undo is the twin of Redo. --> https://github.com/JabRef/jabref/blob/8412e651b671a427c828924ce5698eb4c32f269d/src/main/java/org/jabref/gui/undo/UndoAction.java

Okey. Should I send it to the same branch and same pr?

> @subhramit We should test this. This PR implements both suggestions of you ([#11809 (comment)](https://github.com/JabRef/jabref/issues/11809#issuecomment-2367033790)). I think, one needs to open 5 to 10 libraries, switch tabs back and forth, modifying the library and see if undo/redo is enabled properly.
> 
> @melisolmez For consistency reasons, this patch also nees to be ported to the UndoAction. Undo is the twin of Redo. --> https://github.com/JabRef/jabref/blob/8412e651b671a427c828924ce5698eb4c32f269d/src/main/java/org/jabref/gui/undo/UndoAction.java

Okey. Should I send it to the same branch and same pr?

**Comment by melisolmez** [comment]
> @subhramit We should test this. This PR implements both suggestions of you ([#11809 (comment)](https://github.com/JabRef/jabref/issues/11809#issuecomment-2367033790)). I think, one needs to open 5 to 10 libraries, switch tabs back and forth, modifying the library and see if undo/redo is enabled properly.
> 
> @melisolmez For consistency reasons, this patch also nees to be ported to the UndoAction. Undo is the twin of Redo. --> https://github.com/JabRef/jabref/blob/8412e651b671a427c828924ce5698eb4c32f269d/src/main/java/org/jabref/gui/undo/UndoAction.java

Okey. Should I send it to the same branch?

**Comment by subhramit** [comment]
> @subhramit We should test this. This PR implements both suggestions of you ([#11809 (comment)](https://github.com/JabRef/jabref/issues/11809#issuecomment-2367033790)). I think, one needs to open 5 to 10 libraries, switch tabs back and forth, modifying the library and see if undo/redo is enabled properly.

Actually, suggestion 1 was to conditionally but explicitly use `removeListener`, which is not what has been done in the PR. `WeakChangeListner` automatically takes care of that. 

What is probably confusing - explicit unbinding - I think not really necessary here (I dont remember how the garbage collecter deals with a chain of references), but still a safe practice as there is a life cycle mismatch between RedoAction and tabs - RedoAction might live longer than the relevance of a particular tab. There can be "zombie" bindings between `executable` and the tab. By unbinding, the old/irrelevant tabs can be garbage collected even if `RedoAction` itself is still alive.

File: src/main/java/org/jabref/gui/undo/RedoAction.java [hunk 4]
```diff
@@ -21,11 +25,15 @@ public RedoAction(Supplier<LibraryTab> tabSupplier, DialogService dialogService,
...
+            oldValue.ifPresent(libraryTab -> this.executable.unbind());
```
Comment (subhramit) [inline src/main/java/org/jabref/gui/undo/RedoAction.java:32]: Thinking a bit more, I am a bit concerned about the order of operations here.
Case 1:
State A->B - change detected, bind to B, unbind if oldValue is not null. After this,
State B->A - change detected, bind to A (not needed), unbind if B (oldvalue) is not null, which is true, so unbind (this is fine).

Case 2:
State A->B - change detected, bind to B, unbind if oldValue is not null. After this,
State B->C - change detected, bind to C, unbind if B is not null (which is true) - thus executable is not bound.

@Siedlerchr would it be better if we don't unbind at all here? Or is this alright as LibraryTab can never have a third state "C"?

**Comment by HoussemNasri** [comment]
> Okey. Should I send it to the same branch?

Yes, push the changes to this branch and it should become visible on GitHub.

**Comment by melisolmez** [comment]
@HoussemNasri hello, Could you review the pr?

**Comment by Siedlerchr** [comment]
We have tested this with the following test cases:

1. Create a couple of new libraries
2. In Each one enter a new article (+ button)  with author a in the first, b in the second and so on 
3. Do Redo and switch tabs 
grafik

**Comment by subhramit** [comment]
> We have tested this with the following test cases and found some odd issues:
> 
> 1. Create a couple of new libraries
> 2. In Each one enter a new article (+ button)  with author a in the first, b in the second and so on
> 3. Do Undo, switch tabs, Redo, etc.

More details: Compare it with the main branch. You will find that on adding the author name, it is not reflected in the author column of the entry (until some sequence of events - changing tabs, pressing enter, etc.)
Furthermore, the undo/redo changes are not reflected without switching tabs.

My hypothesis: the early unbinding (described in https://github.com/JabRef/jabref/pull/11839#discussion_r1778250940) is causing nothing to be bound, and thus no listeners are attached at all, and thus no changes are reflected. 

Fix: The unbinding should not happen where it is currently. Remove it, profile the memory using [VirtualVM](https://visualvm.github.io/download.html) to test the changes. We profiled it on the main branch - on adding tabs, the used heap size was shooting up, whereas in your branch the garbage collector did kick in, reducing the heap size on closing tabs. 

If the sequence becomes too confusing, you can also use plain old `addListener` and `removeListener` conditionally.

**Comment by melisolmez** [comment]
> > We have tested this with the following test cases and found some odd issues:
> > 
> > 1. Create a couple of new libraries
> > 2. In Each one enter a new article (+ button)  with author a in the first, b in the second and so on
> > 3. Do Undo, switch tabs, Redo, etc.
> 
> More details: Compare it with the main branch. You will find that in your branch, on adding the author name, it is not reflected in the Author/Editor column of the entry (until some sequence of events - changing tabs, pressing enter, etc.) Furthermore, the undo/redo changes are not reflected without switching tabs.
> 
> My hypothesis: the early unbinding (described in [#11839 (comment)](https://github.com/JabRef/jabref/pull/11839#discussion_r1778250940)) is causing nothing to be bound, and thus no listeners are attached at all, and thus no changes are reflected.
> 
> Fix: The unbinding should not happen where it is currently. Remove it, profile the memory using [VirtualVM](https://visualvm.github.io/download.html) to test the changes. We profiled it on the main branch - on adding tabs, the used heap size was shooting up, whereas in your branch the garbage collector did kick in, reducing the heap size on closing tabs.
> 
> If the sequence becomes too confusing, you can also use plain old `addListener` and `removeListener` conditionally.
> 
> Apologies that the testing part may make this a bit more involved than a good first issue.

Thank you for your feedbak. I will look at it in more detail

**Comment by melisolmez** [comment]
Hi @subhramit, I tested what you said but I couldn't find where it was wrong. When I add author name it is reflected in the  Author/Editor column. and the undo and redo buttons active. 

I don't understand exactly what the problem is, can you explain it in more detail?

**Comment by subhramit** [comment]
> I don't understand exactly what the problem is, can you explain it in more detail?

Will try to send you a video.

**Comment by subhramit** [comment]
I am unable to reproduce what @Siedlerchr did, the undo/redo buttons are staying disabled for me:
image
The undo button becomes active as soon as I type "a" as the author of a new entry in the `main` branch.

**Comment by subhramit** [comment]
In the above case, I closed some previously opened libraries.
Here is another case (no libraries open on starting JabRef):
Add a library:
image

Both the Undo/Redo buttons are active even though no changes have been made/ no entries have been added.
Comparison with `main` branch: On opening JabRef with no previous libraries, the Undo/Redo buttons are enabled (because of the other bug), but as soon as a new library is added, the buttons become disabled. When A new entry is added to the library, the Undo button becomes active.
Conclusion: the listener in `main` is capturing changes properly, the one in this branch is not.

**Comment by subhramit** [comment]
The two issues mentioned above persist even on removing the line that unbinds the executable.

**Comment by subhramit** [comment]
@melisolmez I think the other maintainers are presently a bit busy, and we appreciate your patience in sticking to the task.
What I think you can do is meet the two conditions - 
1. resolve the issues so that the unexpected behaviour above is handled (I suggest that you use the old logic and conditionally remove the listeners using an "else" to the `ifPresent` check for a tab), and 
2. confirm with VirtualVM after closing tabs that the listeners are indeed cleaned up after that change.

(Optional) I think testing can be less confusing if https://github.com/JabRef/jabref/issues/11837 is solved first, you can also pick it up if interested, or continue with this one.

**Comment by melisolmez** [comment]
Hello @subhramit, I tested your sayng but I couldn't find the wrong things where. 

> @melisolmez I think the other maintainers are presently a bit busy, and we appreciate your patience in sticking to the task. What I think you can do is meet the two conditions -
> 
> 1. resolve the two issues mentioned above so that the unexpected behaviour above is handled (I suggest that you use the old logic and conditionally remove the listeners using an "else" to the `ifPresent` check for a tab), and
> 2. confirm with VirtualVM after closing tabs that the listeners are indeed cleaned up after that change.
> 
> I think testing can be less confusing if #11837 is solved first, but that is optional - you can also pick it up if interested, or continue with this one. 

Hello @subhramit thank you for help. I understand what is the problem and I will try fix it. 
so I need a few details to solve this problem. I would be glad if you answer.

1. How to manage active and passive buttons? and where ?( for example in the main branch: when open a library these buttons can be disable. If I find where this event where it can help me with issue)
2. If I examine where and how it is managed, I will try to establish a similar structure, so I will look for a solution to issue #11837.

**Comment by subhramit** [comment]
> 1. How to manage active and passive buttons? and where ?( for example in the main branch: when open a library these buttons can be disable. If I find this event where it can help me with issue)
> 2. If I examine where and how it is managed, I will try to establish a similar structure, so I will look for a solution to issue [redo and undo buttons are active without open library  #11837](https://github.com/JabRef/jabref/issues/11837).

Good questions.
I have not worked on this component, so @HoussemNasri would be able to give you more context. But I'll tell you how I would investigate this:
If we look at https://github.com/JabRef/jabref/pull/11758/files, we can see these that set the property undoable and redoable:
https://github.com/JabRef/jabref/blob/4b8d66863cdbb9d54bea569ab36ed4b1b3498cd0/src/main/java/org/jabref/gui/undo/CountingUndoManager.java#L27-L31

Thus, I would start at `UndoManager.java` (it contains logic for setting both undo and redo buttons).

**Comment by subhramit** [comment]
> Thus, I would start at ~`UndoManager.java`~ `CountingUndoManager.java`(it contains logic for setting both undo and redo buttons).

More thoughts:
When no library is open (I think that translates to no open databases)
```Java
if (databases == null || databases.isEmpty()) {
```
Then the properties should be set to disabled.

**Comment by subhramit** [comment]
> ```java
> if (databases == null || databases.isEmpty())
> ```

@Siedlerchr can we use Injector to get `BibDatabase` here?

**Comment by Siedlerchr** [comment]
one liner 
`         this.executable.bind(needsDatabase(stateManager)`

**Comment by HoussemNasri** [comment]
Hello @melisolmez,

Thank you for your hard work in addressing this issue. The Undo/Redo feature is indeed quite challenging, especially given its current state, which requires significant refactoring.

We'll take over from here and continue refining the pull request to bring it to a mergeable state.

In the meantime, feel free to explore other good first issues and comment on any that interest you and someone will assign it to you.

**Review by HoussemNasri — APPROVED** [review]
Thank you again for your work! We truly appreciate your efforts and look forward to more contributions.

**Review by Siedlerchr — APPROVED** [review]

**Pull request merged**